File: examples3d/heightfield3.rs
use na::{ComplexField, DMatrix, Isometry3, Point3, Vector3};
use rapier3d::dynamics::{JointSet, RigidBodyBuilder, RigidBodySet};
use rapier3d::geometry::{ColliderBuilder, ColliderSet, SharedShape};
use rapier_testbed3d::Testbed;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let joints = JointSet::new();

    /*
     * Ground
     */
    let ground_size = Vector3::new(100.0, 1.0, 100.0);
    let nsubdivs = 20;

    let heights = DMatrix::from_fn(nsubdivs + 1, nsubdivs + 1, |i, j| {
        if i == 0 || i == nsubdivs || j == 0 || j == nsubdivs {
            10.0
        } else {
            let x = i as f32 * ground_size.x / (nsubdivs as f32);
            let z = j as f32 * ground_size.z / (nsubdivs as f32);

            // NOTE: make sure we use the sin/cos from simba to ensure
            // cross-platform determinism of the example when the
            // enhanced_determinism feature is enabled.
            ComplexField::sin(x) + ComplexField::cos(z)
        }
    });

    let rigid_body = RigidBodyBuilder::new_static().build();
    let handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::heightfield(heights, ground_size).build();
    colliders.insert(collider, handle, &mut bodies);

    /*
     * Create the cubes
     */
    let num = 8;
    let rad = 1.0;

    let shift = rad * 2.0 + rad;
    let centerx = shift * (num / 2) as f32;
    let centery = shift / 2.0;
    let centerz = shift * (num / 2) as f32;

    for j in 0usize..20 {
        for i in 0..num {
            for k in 0usize..num {
                let x = i as f32 * shift - centerx;
                let y = j as f32 * shift + centery + 3.0;
                let z = k as f32 * shift - centerz;

                // Build the rigid body.
                let rigid_body = RigidBodyBuilder::new_dynamic().translation(x, y, z).build();
                let handle = bodies.insert(rigid_body);

                let collider = match j % 6 {
                    0 => ColliderBuilder::cuboid(rad, rad, rad).build(),
                    1 => ColliderBuilder::ball(rad).build(),
                    // Rounded cylinders are much more efficient that cylinder, even if the
                    // rounding margin is small.
                    2 => ColliderBuilder::round_cylinder(rad, rad, rad / 10.0).build(),
                    3 => ColliderBuilder::cone(rad, rad).build(),
                    4 => ColliderBuilder::capsule_y(rad, rad).build(),
                    _ => {
                        let shapes = vec![
                            (
                                Isometry3::identity(),
                                SharedShape::cuboid(rad, rad / 2.0, rad / 2.0),
                            ),
                            (
                                Isometry3::translation(rad, 0.0, 0.0),
                                SharedShape::cuboid(rad / 2.0, rad, rad / 2.0),
                            ),
                            (
                                Isometry3::translation(-rad, 0.0, 0.0),
                                SharedShape::cuboid(rad / 2.0, rad, rad / 2.0),
                            ),
                        ];

                        ColliderBuilder::compound(shapes).build()
                    }
                };

                colliders.insert(collider, handle, &mut bodies);
            }
        }
    }

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, joints);
    testbed.look_at(Point3::new(100.0, 100.0, 100.0), Point3::origin());
}

fn main() {
    let testbed = Testbed::from_builders(0, vec![("Boxes", init_world)]);
    testbed.run()
}

File: bindings/python/rapier-testbed/rapier_testbed/examples3/heightfield3.py
"""Port of examples3d/heightfield3.rs."""
from __future__ import annotations

import math

import numpy as np

import rapier3d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "Heightfield"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = (100.0, 1.0, 100.0)
    nsubdivs = 20
    heights = np.zeros((nsubdivs + 1, nsubdivs + 1), dtype=np.float32)
    for i in range(nsubdivs + 1):
        for j in range(nsubdivs + 1):
            if i == 0 or i == nsubdivs or j == 0 or j == nsubdivs:
                heights[i, j] = 10.0
            else:
                x = i * ground_size[0] / nsubdivs
                z = j * ground_size[2] / nsubdivs
                heights[i, j] = math.sin(x) + math.cos(z)

    ground = rp.RigidBody.fixed()
    ground_h = bodies.insert(ground)
    colliders.insert_with_parent(
        rp.Collider.heightfield(heights, ground_size), ground_h, bodies
    )

    num = 8
    rad = 1.0
    shift = rad * 2.0 + rad
    centerx = shift * (num // 2)
    centery = shift / 2.0
    centerz = shift * (num // 2)
    for j in range(20):
        for i in range(num):
            for k in range(num):
                x = i * shift - centerx
                y = j * shift + centery + 3.0
                z = k * shift - centerz
                body = rp.RigidBody.dynamic().translation((x, y, z))
                h = bodies.insert(body)
                kind = j % 6
                if kind == 0:
                    col = rp.Collider.cuboid(rad, rad, rad)
                elif kind == 1:
                    col = rp.Collider.ball(rad)
                elif kind == 2:
                    col = rp.Collider.round_cylinder(rad, rad, rad / 10.0)
                elif kind == 3:
                    col = rp.Collider.cone(rad, rad)
                elif kind == 4:
                    col = rp.Collider.capsule_y(rad, rad)
                else:
                    shapes = [
                        (
                            rp.Isometry3.identity(),
                            rp.SharedShape.cuboid(rad, rad / 2.0, rad / 2.0),
                        ),
                        (
                            rp.Isometry3.from_translation(rad, 0.0, 0.0),
                            rp.SharedShape.cuboid(rad / 2.0, rad, rad / 2.0),
                        ),
                        (
                            rp.Isometry3.from_translation(-rad, 0.0, 0.0),
                            rp.SharedShape.cuboid(rad / 2.0, rad, rad / 2.0),
                        ),
                    ]
                    col = rp.Collider.compound(shapes)
                colliders.insert_with_parent(col, h, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((100.0, 100.0, 100.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
